### PR TITLE
Misc. additions and extensions

### PR DESCRIPTION
**Pull request by prburke**
Misc. fixes and additions (some moved from SD repository).

Tests required.  Might do that for hack days :wink: 

@emilienoel @jeromecornet

### CODE REVIEW

**Comment by prburke** [comment]
@emilienoel

File: lib/netsuite/actions/async_add_list.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+          response_hash = @response.to_hash
+          response_hash[:async_add_list_response] ? response_hash[:async_add_list_response][:async_status_result] : nil
+        end
+      end
+
+      def success?
+        !response_body.nil?
+      end
+
+      module Support
```
Comment (emilienoel) [inline lib/netsuite/actions/async_add_list.rb:63]: So you can't use `ActiveSupport::` at all? http://api.rubyonrails.org/classes/ActiveSupport/Concern.html would be good here...

File: lib/netsuite/async_result.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      errors = []
+
+      if async_result[:read_response_list]
+        response_list = async_result[:read_response_list][:read_response]
+      else
+        response_list = async_result[:write_response_list][:write_response]
+      end
+
+      Array[response_list].flatten.each do |response|
+        if response[:status][:@is_success] == 'true'
```
Comment (emilienoel) [inline lib/netsuite/async_result.rb:20]: Nitpick but could do `!!response[:status][:@is_success]` instead.

File: lib/netsuite/async_result.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        elsif response[:status][:status_detail]
+           response_errors = Array[response[:status][:status_detail]].flatten.map do |details|
+             NetSuite::Error.new(details)
+           end
+           errors << [response[:base_ref][:@external_id], response_errors]
+        end
+      end
+
+      @base_ref_list = NetSuite::Records::BaseRefList.new(base_ref: base_ref_list)
+      @errors = Hash[errors]
```
Comment (emilienoel) [inline lib/netsuite/async_result.rb:31]: How about `@errors = {}`...  `@errors[response[:base_ref][:@external_id]]` = response_errors

This way you can get rid of `errors = []`

File: lib/netsuite/actions/async_add_list.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      #</soapenv:Body>
+      def response_body
+        @response_body ||= begin
+          response_hash = @response.to_hash
+          response_hash[:async_add_list_response] ? response_hash[:async_add_list_response][:async_status_result] : nil
+        end
+      end
+
+      def success?
+        !response_body.nil?
```
Comment (emilienoel) [inline lib/netsuite/actions/async_add_list.rb:60]: Could this be returning something like 'error' in the response_body? Would this be considered as a success?

File: lib/netsuite/records/vendor_bill.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+module NetSuite
+  module Records
+    class VendorBill
+      include Support::Fields
+      include Support::RecordRefs
+      include Support::Records
+      include Support::Actions
+      include Namespaces::TranPurch
+
+      actions :add, :delete, :get, :get_list, :get_select_value, :initialize, :search, :update, :upsert, :upsert_list
```
Comment (emilienoel) [inline lib/netsuite/records/vendor_bill.rb:10]: Are the `async` options available here too?

File: lib/netsuite/actions/async_add_list.rb (hunk 1 above)
Comment (jeromecornet) [inline lib/netsuite/actions/async_add_list.rb:63]: the problem is that you need rails, which is good for us but bad for people wanting to just use the gem in ruby projects

File: lib/netsuite/actions/async_add_list.rb (hunk 4 above)
Comment (prburke) [inline lib/netsuite/actions/async_add_list.rb:60]: If the response_hash contains an async_status_result (line 55) then we consider it a success.  If we get this far, the job finished running.  The async_status_result itself can contain errors mind you (some of the records that the caller attempted to add might have been invalid).  If that is the case the caller would have been notified in advance because the AsyncStatus for the job would have been 'finishedWithErrors': https://github.com/prburke/netsuite/blob/initial_extensions/lib/netsuite/async_status.rb#L41
If that's the case then the errors can be extracted via the AsyncResult: https://github.com/prburke/netsuite/blob/initial_extensions/lib/netsuite/async_result.rb#L5
I have lots of ideas on how this async api code can be improved so the overall flow will likely change.

File: lib/netsuite/records/vendor_bill.rb (hunk 5 above)
Comment (prburke) [inline lib/netsuite/records/vendor_bill.rb:10]: Ooops...I thought I added them.  I've only tested synchronous imports so far.

File: lib/netsuite/async_result.rb (hunk 2 above)
Comment (jeromecornet) [inline lib/netsuite/async_result.rb:20]: nah, 'false' is truthy

File: lib/netsuite/async_result.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      base_ref_list = []
+      errors = []
+
+      if async_result[:read_response_list]
+        response_list = async_result[:read_response_list][:read_response]
+      else
+        response_list = async_result[:write_response_list][:write_response]
+      end
+
+      Array[response_list].flatten.each do |response|
```
Comment (jeromecornet) [inline lib/netsuite/async_result.rb:19]: very nitpicky, but a more ruby way to do this would be to do

```
responses=Array[response_list].flatten
base_ref_list = responses.select { |response| response[:status][:@is_success] == 'true' }.map { |response| response[:base_ref] }

errors = responses.reject { |response| response[:status][:@is_success] == 'true' }.map do |response| 
  details = Array[response[:status][:status_detail]].flatten.map { |details|  NetSuite::Error.new(details) }
  [response[:base_ref][:@external_id], details]
end
```

Comment (prburke) [inline lib/netsuite/async_result.rb:19]: :smile:  I like it.

File: lib/netsuite/support/actions.rb [hunk 7]
```diff
@@ -28,8 +28,6 @@ def action(name)
             self.send(:include, NetSuite::Actions::GetSelectValue::Support)
           when :search
             self.send(:include, NetSuite::Actions::Search::Support)
-          when :search_more_with_id
```
Comment (jeromecornet) [inline lib/netsuite/support/actions.rb:31]: why did you remove it  here and above ?

**Comment by jeromecornet** [comment]
:+1: 
I hate SOAP

File: lib/netsuite/support/actions.rb (hunk 7 above)
Comment (prburke) [inline lib/netsuite/support/actions.rb:31]: This is a bug in the gem - broken master.  There is a PR in the main repo
to do the same.  Someone deleted the SearchMoreWithId action but didn't
clean everything up properly.

Comment (jeromecornet) [inline lib/netsuite/support/actions.rb:31]: oh ok :-)

File: lib/netsuite/async_result.rb (hunk 2 above)
Comment (emilienoel) [inline lib/netsuite/async_result.rb:20]: right. What's the point of using !! then? I thought it was used in this case. Wow, fail on my part. Good catch @jeromecornet

**Comment by emilienoel** [comment]
LGTM!

File: lib/netsuite/async_result.rb (hunk 3 above)
Comment (prburke) [inline lib/netsuite/async_result.rb:31]: Sounds good.  I'm going to incl this change in a different PR.
